### DISCUSSION TITLE
Interesting result, too good for true !

### DISCUSSION THREAD

**Discussion by turkalpmd** [pycaret/pycaret#3223] [Q&A]
Why pycaret.regression(transformation=True) transforming to dummies?

Because it is not true, I have already described to categoric attributes, I don't need to transformation. How can I solve this?

**Comment by turkalpmd** [comment]
Dear @ngupta23 & @moezali1 ;

I am seeing unusual results when setting "max_encoding_ohe=0" in my regression model. The model is achieving unreasonably high performance, but the validation residuals are very poor. However, when I tested the model on data that was not used in training, the results were excellent. This is important to me because I plan to use the model in an academic publication. Can you help me understand how the model is able to achieve such good results on the test data through encoding categorical variables, despite the poor validation performance?


Here is the "max_encoding_ohe=0" result;
Train and validation:
![image](https://user-images.githubusercontent.com/85236337/210564208-24bc526a-cf6a-4897-ac33-b9d4b9bdfc59.png)

Test:
Extreme Gradient Boosting | MAE:  0.1528 | MSE: 0.1250 | RMSE: 0.3536 | R2: 0.9364 | RMSLE: 0.0280 | MAPE: 0.0136

Here is the "max_encoding_ohe=Default" result;
Train and test:
![image](https://user-images.githubusercontent.com/85236337/210564775-f3f62f75-5244-4f7b-9144-2fa2c322f040.png)


Test: 

Extreme Gradient Boosting | MAE:  1.0108 | MSE: 1.6278 | RMSE: 1.2759 | R2: 0.1721 | RMSLE: 0.1081 | MAPE: 0.0896


Sinerely,

İzzet

**Comment by ngupta23** [comment]
@tvdboom - FYI...

**Comment by tvdboom** [comment]
@turkalpmd Could you share a minimal reproducible example? (data + code) We can't debug without.

**Reply by TarikKaanKoc** [reply]
@turkalpmd  ++

**Comment by turkalpmd** [comment]
Here is the my data and code;

[pycaret_issue.zip](https://github.com/pycaret/pycaret/files/10355116/pycaret_issue.zip)


I have split the train and test data and provided only the train data to the model. The data I am using is real, and the model's performance is unusual. I do not believe there is data leakage, as a single variable in the setup file cannot explain such a strong result. I am unsure of what is causing this. While I am pleased with the excellent result, I am frustrated by my lack of understanding. I would like to report this amazing outcome, but I want to understand the cause of my frustration. Thank you for your contribute.

**Comment by tvdboom** [comment]
I had a look. Not sure what's happening but I wouldn't trust the results tbh. I tried different encoding methods and they all perform poorly except LeaveOneOut, which is the default encoder when you don't apply OneHot. Something fishy is going on, especially since TargetEncoder, which is very similar to LeaveOneOut also yields poor results. Not sure if it's leakage or a bug in the Encoder... Any ideas @ngupta23 @Yard1 ?

**Reply by Yard1** [reply]
In other words, we suspect the issue is with LeaveOneOut?

**Reply by turkalpmd** [reply]
I was using global variables named train and test in my code, and I started to wonder if there might be a data leak because these variables are also defined in the source code. As a precaution, I decided to change the names of these variables to train_ and test_, but to my surprise, this did not affect the accuracy of my model. In fact, it is still achieving a very high accuracy. I am now thinking about taking a look at the source code to try to understand this behavior.

**Reply by turkalpmd** [reply]
I think that maybe leaving one out encoding is the best way to go, so I implemented it with low code, but the results were still bad. It's possible that there is a bug causing this.

[low_code.ipynb.zip](https://github.com/pycaret/pycaret/files/10356324/low_code.ipynb.zip)

**Reply by TarikKaanKoc** [reply]
> I think that maybe leaving one out encoding is the best way to go, so I implemented it with low code, but the results were still bad. It's possible that there is a bug causing this.
> 
> [low_code.ipynb.zip](https://github.com/pycaret/pycaret/files/10356324/low_code.ipynb.zip)

It is unlikely that using global variables with the same names as those used in the source code would cause a data leak or affect the accuracy of the model. 

If you are still seeing poor performance on the validation data, it is possible that there is a problem with the data or the model. It could be helpful to try different models or different encoding methods to see if the performance improves. It may also be helpful to visualize the data and the residuals to get a better understanding of the problem.

**I'll take a look at the .ipynb file you sent.**

**Reply by turkalpmd** [reply]
Thank you Kaan, I will be looking forward to your solutions.

**Reply by TarikKaanKoc** [reply]
I do not see any syntax errors or other issues in the code that you have provided. However, without more information about the data and the specific problem you are trying to solve, it is difficult to say whether the code will produce good results or whether there are any issues that may affect the model's performance.

It could be that the model is not a good fit for the data, or that the features are not providing enough information to make accurate predictions. It could also be that the data has noise or outliers that are affecting the model's performance.

To try to improve the model's performance, you could try adding or removing features, or trying different models and hyperparameter values. It may also be helpful to try different encoding methods or to transform the data in some way to make it more suitable for modeling. 

### Review the data and the following steps:

- data cleaning 
- feature selection and feature engineering  
- encoding methods and transform
- models and hyperparameter

**Reply by TarikKaanKoc** [reply]
#### To prevent overfitting and improve the model's performance, you can try a number of different approaches: 

* Increase the size of the training data: This can help the model learn more general patterns and improve its ability to generalize to new data.
* Use cross-validation: This can help you get a better estimate of the model's performance and identify any issues with overfitting.
* Try different models: Different models may have different properties and may be better suited for different types of data. It may be helpful to try multiple models and compare the results.
* Visualize the data and the residuals: This can help you identify any issues with the data or the model, and may suggest ways to improve the model's performance. (ex: see quantiles or use boxplot, z-score etc)

**Comment by TarikKaanKoc** [comment]
Without more information about the data and the specific steps you've taken, we're unlikely to be able to offer specific recommendations for your model... (My ideas will be conjecture.) 

Setting the max_encoding_ohe parameter to zero will disable one-hot encoding for the categorical variables and use a different encoding method instead. 

If you are seeing poor performance on the validation data and good performance on the test data after setting max_encoding_ohe to zero, it is possible that the model is still overfitting to the training data. In this case, you could try using a different encoding method or implementing regularization techniques to help prevent overfitting. It may also be helpful to visualize the residuals and plot the predicted values against the actual values to get a better understanding of the model's performance.

**Reply by turkalpmd** [reply]
Sure, my last opinion is I found a BUG from source code 😄 I must try other algorithms maybe my data have got vanishing gradients not sure.
